src/Ledger/Chain.lagda

```diff
@@ -89,12 +90,51 @@ data _⊢_⇀⦇_,NEWEPOCH⦈_ : NewEpochEnv → NewEpochState → Epoch → New
 \end{figure*}
 
 \begin{code}[hide]
+maybePurpose : DepositPurpose → (DepositPurpose × Credential) → Coin → Maybe Coin
+maybePurpose prps (prps' , _) c with prps ≟ prps'
+... | yes _ = just c
+... | no _ = nothing
+
+maybePurpose-prop : ∀ {prps} {x} {y}
+  → (m : (DepositPurpose × Credential) ↛ Coin)
+  → (x , y) ∈ dom ((mapMaybeWithKeyᵐ (maybePurpose prps) m) ˢ)
+  → x ≡ prps
+maybePurpose-prop {prps = prps} {x} {y} _ xy∈dom with to dom∈ xy∈dom
+... | z , ∈mmwk with prps ≟ x | ∈-mapMaybeWithKey {f = maybePurpose prps} ∈mmwk
+... | yes refl | _ = refl
+
+filterPurpose : DepositPurpose → (DepositPurpose × Credential) ↛ Coin → Credential ↛ Coin
+filterPurpose prps m = mapKeys proj₂ (mapMaybeWithKeyᵐ (maybePurpose prps) m) λ where
+  {x , .z} {y , z} x∈dom y∈dom refl → case maybePurpose-prop {prps = prps} m x∈dom of λ where
+    x≡prps → case maybePurpose-prop {prps = prps} m y∈dom of λ where
+      refl → cong (_, z) x≡prps
+
+instance
+  _ = +-0-monoid
+
+calculateStakeDistrs : LState → StakeDistrs
+calculateStakeDistrs ls =
+  let open LState ls; open CertState certState; open PState pState; open UTxOState utxoSt
+      spoDelegs = ∅ᵐ -- TODO
+      drepDelegs = ∅ᵐ -- TODO
+      govActionDeposits = foldl _∪⁺_ ∅ᵐ $ setToList $
+        map
+          (λ where record { returnAddr = record {stake = c} ; deposit = v } → ❴ credVoter DRep c , v ❵ᵐ)
```

Comment:
This is still technically incorrect, since you get the `VDeleg` by looking up `c` in `voteDelegs` from `DState`. If you want you can fix it now or in a later PR.